validator/project_validator.go

```diff
@@ -1850,178 +1804,68 @@ func validateTaskSyncCommands(p *model.Project, runLong bool) ValidationErrors {
+// The dependedOnTask and every other task along the path must run on all the same requester types as the dependentTask
+// and the dependency on the dependedOnTask must be with a status in statuses, if provided.
+func validateTVDependsOnTV(dependentTask, dependedOnTask model.TVPair, statuses []string, project *model.Project) error {
+	g := project.DependencyGraph()
+	tvTaskUnitMap := tvToTaskUnit(project)
 
-	sourceTask, ok := tvToTaskUnit[source]
-	if !ok {
-		return errors.Errorf("could not find task '%s' in build variant '%s'",
-			source.TaskName, source.Variant)
-	}
+	dependentNode := task.TaskNode{Name: dependentTask.TaskName, Variant: dependentTask.Variant}
+	dependedOnNode := task.TaskNode{Name: dependedOnTask.TaskName, Variant: dependedOnTask.Variant}
+	traversal := func(edge task.DependencyEdge) bool {
+		dependent := edge.From
+		dependedOn := edge.To
 
-	// patches and mainline builds shouldn't depend on anything that's git tag only,
-	// while something that could run in a git tag build can't depend on something that's patchOnly.
-	// requireOnNonGitTag is just requireOnPatches & requireOnNonPatches so we don't consider this case.
-	depReqs := dependencyRequirements{
-		lastDepNeedsSuccess: true,
-		requireOnPatches:    !sourceTask.SkipOnPatchBuild() && !sourceTask.SkipOnNonGitTagBuild(),
-		requireOnNonPatches: !sourceTask.SkipOnNonPatchBuild() && !sourceTask.SkipOnNonGitTagBuild(),
-		requireOnGitTag:     !sourceTask.SkipOnNonPatchBuild() && !sourceTask.SkipOnGitTagBuild(),
-	}
-	depFound, err := dependencyMustRun(target, source, depReqs, allTVs, visited, tvToTaskUnit)
-	if err != nil {
-		return errors.Wrapf(err, "error searching for dependency of task '%s' in build variant '%s'"+
-			" on task '%s' in build variant '%s'",
-			source.TaskName, source.Variant,
-			target.TaskName, target.Variant)
-	}
-	if !depFound {
-		errMsg := "task '%s' on build variant '%s' must depend on" +
-			" task '%s' in build variant '%s' running and succeeding"
-		if depReqs.requireOnPatches && depReqs.requireOnNonPatches {
-			errMsg += " for both patches and non-patches"
-		} else if depReqs.requireOnPatches {
-			errMsg += " for patches"
-		} else if depReqs.requireOnNonPatches {
-			errMsg += " for non-patches"
-		} else if depReqs.requireOnGitTag {
-			errMsg += " for git-tag builds"
-		}
-		return errors.Errorf(errMsg, source.TaskName, source.Variant, target.TaskName, target.Variant)
-	}
-	return nil
-}
+		dependentTaskUnit := tvTaskUnitMap[model.TVPair{TaskName: dependent.Name, Variant: dependent.Variant}]
+		dependedOnTaskUnit := tvTaskUnitMap[model.TVPair{TaskName: dependedOn.Name, Variant: dependedOn.Variant}]
 
-type dependencyRequirements struct {
-	lastDepNeedsSuccess bool
-	requireOnPatches    bool
-	requireOnNonPatches bool
-	requireOnGitTag     bool
-}
+		var dep model.TaskUnitDependency
+		for _, dependency := range dependentTaskUnit.DependsOn {
+			if dependency.Name == dependedOn.Name && dependency.Variant == dependedOn.Variant {
```

Comment:
It's a bit confusing for this to be called only "dep" when we have both dependent and dependedOn being used; can we call this something like dependentInfo? (just a suggestion, we can also leave it but a comment would be helpful)

Reply:
I renamed it to `edgeInfo`